README.md
# chkplot
[![npm version](https://badge.fury.io/js/chkplot.svg)](https://badge.fury.io/js/chkplot) [![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)

Chia plot log parser/analyzer

## Install
```
npm install chkplot
# or
yarn add chkplot
```

## Command line
```
npx chkplot list [-n <N>] [-c]
    Show available plot log files.
    -n: If you specify -n 3, then top 3 of most recent started plotting progress/result will be shown.
    -c: Compact output

npx chkplot wip [-n <N>] [-c]
    Show plotting progress from plotter log files.
    -n: If you specify -n 10, then top 10 of most recent started plotting progress will be shown.
    -c: Compact output

npx chkplot summary [-u <uuid>|-n <N>|-a]
    Show plot summary.
    -u: Specify plot uuid for summary. uuid can be listed by 'npx chkplot list'
    -n: If you specify -n 3, then top 3 of most recent started plotting log summary will be shown.
    -a: Show all available plot log summary

npx chkplot watch
    Realtime monitor for plot progress.
```
* If you globally install `chkplot` such as `npm install -g chkplot` or `yarn add global chkplot`,  
  you can directly run above command without `npx` command prefix.  
  
  **Example:**
  ```shell
  chkplot wip -n 10
  ```

### chkplot list
```shell
$ npx chkplot list -n 4

npx: installed 1 in 1.735s
5f4f50c2-c4f4-441c-be73-8d72af6342f6 31.3% phase2 20210604_123039 k:32 r:2 b:4600MiB t:E:\chia_plot d:?
a065f74d-cf80-4784-99a1-627dd58fac35 31.4% phase2 20210604_123039 k:32 r:2 b:4600MiB t:D:\chia_plot d:?
e613f9c7-dfc7-4e4d-8829-15945c73e7f3 100%  complete 20210604_003946 k:32 r:4 b:4600MiB t:E:\chia_plot d:?
eba1de83-12be-4c7e-8936-c92da581e4d2 100%  complete 20210604_003946 k:32 r:4 b:4600MiB t:D:\chia_plot d:?
```
```shell
$ npx chkplot list -n 4 -c

npx: installed 1 in 1.735s
5f4f50c2 31.4% phase2 20210604_123039 k:32 r:2 b:4600MiB
a065f74d 31.4% phase2 20210604_123039 k:32 r:2 b:4600MiB
e613f9c7 100%  complete 20210604_003946 k:32 r:4 b:4600MiB
eba1de83 100%  complete 20210604_003946 k:32 r:4 b:4600MiB
```

### chkplot wip
Show only work in progress plotting tasks
```shell
$ npx chkplot wip

npx: installed 1 in 1.735s
5f4f50c2-c4f4-441c-be73-8d72af6342f6 31.4% phase2 20210604_123039 k:32 r:2 b:4600MiB t:E:\chia_plot d:?
a065f74d-cf80-4784-99a1-627dd58fac35 31.4% phase2 20210604_123039 k:32 r:2 b:4600MiB t:D:\chia_plot d:?
```
```shell
$ npx chkplot wip -c

npx: installed 1 in 1.735s
5f4f50c2 31.4% phase2 20210604_123039 k:32 r:2 b:4600MiB
a065f74d 31.4% phase2 20210604_123039 k:32 r:2 b:4600MiB
```

### chkplot summary
```shell
$ npx chkplot summary -u e613f9c7

npx: installed 1 in 1.735s
uuid                : e613f9c7-dfc7-4e4d-8829-15945c73e7f3
id                  : f8a891124b31847bcc3dceaae7ab740ec50ebbf659520a1e335d47e860106772
start_date          : 20210604_003946
k                   : 32
r                   : 4
b                   : 4600MiB
t                   : E:\chia_plot
d                   :
phase1CompleteTime  : 01:41:41
phase2CompleteTime  : 00:52:52
phase3CompleteTime  : 01:37:37
phase4CompleteTime  : 00:07:07
plotCompleteTime    : 04:18:18
copyTime            : 00:16:16
overallCompleteTime : 04:35:35
progress            : 100
phase               : complete
finish_date         : 2021-6-4 5:14:50
```

### chkplot watch
Realtime plot progress monitor. Press 'q' to exit.
```shell
$ npx chkplot watch
```
![](https://github.com/Chia-Mine/chkplot/blob/v0.0.1/example/chkplot_watch.png)


### Demo
![](https://github.com/Chia-Mine/chkplot/blob/v0.0.1/example/chkplot-demo.gif)

## Library
`chkplot` exposes several function to parse/analyze plotter log.

### parsePlotterLog

```typescript
const {parsePlotterLog} = require("chkplot");
const log = fs.readFileSync(<path_to_plotter_log>, {encoding: "utf-8"});
const parsedLog = parsePlotterLog(log);
```
The content of `parsedLog` looks like:
```json
{
  "params": {
    "start_time": "2021-05-19T01:29:19.301Z",
    "k": 32,
    "pool_public_key": "xxxxxxxxxxxxxx",
    "farmer_public_key": "xxxxxxxxxxxxxx",
    "memo": "xxxxxxxxxxxxxx",
    "temp1_dir": "D:\\chia_plot",
    "temp2_dir": "D:\\chia_plot",
    "id": "xxxxxxxxxxxxxx",
    "buffer_size": [
      4100,
      "MiB"
    ],
    "buckets": 128,
    "threads": 2,
    "stripe_size": 65536
  },
  "phase1": {
    "start_date": "2021-05-19T01:29:19.000Z",
    "table1": {
      "f1_complete": {
        "time": 127.007,
        "cpu": 102.34,
        "date": "2021-05-19T01:31:26.000Z"
      }
    },
    ...
  }
  "phase2": {
    "start_date": "2021-05-19T04:45:39.000Z",
    "table7": {
      "time": 101.065,
      "cpu": 19.12,
      "date": "2021-05-19T04:47:21.000Z"
    },
    ...
  },
  "phase3": {
    "tmp_path": "D:\\\\chia_plot\\\\plot-k32-2021-05-19-10-29-xxxxxxxxxxxxxx.plot.2.tmp",
    "start_date": "2021-05-19T05:55:16.000Z",
    "table1and2": {
      "bucketInfo": [
        ...
      ],
      "firstComputationSummary": {
        "time": 677.709,
        "cpu": 85.55,
        "date": "2021-05-19T06:06:34.000Z"
      },
      "secondComputationSummary": {
        "time": 522.29,
        "cpu": 79.7,
        "date": "2021-05-19T06:15:17.000Z"
      },
      "totalCompressSummary": {
        "time": 1200.39,
        "cpu": 83.01,
        "date": "2021-05-19T06:15:17.000Z"
      }
    },
    ...
  },
  "phase4": {
    "tmp_path": "D:\\\\chia_plot\\\\plot-k32-2021-05-19-10-29-xxxxxxxxxxxxxx.plot.2.tmp",
    "start_date": "2021-05-19T08:16:42.000Z",
    "bucketInfo": [...],
    "P1": "0xfff",
    "P2": "0xaaaaaaaaaa",
    "P3": "0xbbbbbbbbbb",
    "P4": "0xcccccccccc",
    "P5": "0xdddddddddd",
    "P6": "0xeeeeeeeeee",
    "P7": "0xffffffffff",
    "C1": "0x1111111111",
    "C2": "0x2222222222",
    "C3": "0x3333333333",
    "complete": {
      "time": 511.085,
      "cpu": 82.85,
      "date": "2021-05-19T08:25:13.000Z"
    }
  },
  "phaseSummary": {
    "approximate_working_space_used": [
      269.387,
      " GiB"
    ],
    "final_file_size": [
      101.379,
      " GiB"
    ],
    "complete": {
      "time": 24954.324,
      "cpu": 112.88,
      "date": "2021-05-19T08:25:13.000Z"
    },
    "finished_time": "2021-05-19T09:27:58.962Z"
  },
  "copyPhase": {
    "from": "D:\\\\chia_plot\\\\plot-k32-2021-05-19-10-29-xxxxxxxxxxxxxx.plot.2.tmp",
    "to": "R:\\\\plot-k32-2021-05-19-10-29-xxxxxxxxxxxxxx.plot.2.tmp",
    "complete": {
      "time": 3762.214,
      "cpu": 0.83,
      "date": "2021-05-19T09:27:55.000Z"
    },
    "removed_temp2_file": "D:\\\\chia_plot\\\\plot-k32-2021-05-19-10-29-xxxxxxxxxxxxxx.plot.2.tmp",
    "renamed_final_file": {
      "from": "R:\\\\plot-k32-2021-05-19-10-29-xxxxxxxxxxxxxx.plot.2.tmp",
      "to": "R:\\\\plot-k32-2021-05-19-10-29-xxxxxxxxxxxxxx.plot"
    }
  }
}
```

### summarize

```typescript
const {summarize} = require("chkplot");
const log = fs.readFileSync(<path_to_plotter_log>, {encoding: "utf-8"});
const summarizedLog = summarize(log);
```
The content of `summarizedLog` looks like:
```json
{
  "id": "xxxxxxxxxxxxxxxxxxxxxx",
  "start_date": "2021-5-25 2:17:19",
  "k": 32,
  "r": 3,
  "b": [ 4100, "MiB" ],
  "t": "D:\\chia_plot",
  "d": undefined,
  "phase1CompleteTime": { "hour": 3, "minute": 4, "second": 54.753 },
  "phase2CompleteTime": { "hour": 1, "minute": 16, "second": 9.744 },
  "phase3CompleteTime": { "hour": 2, "minute": 33, "second": 14.112 },
  "phase4CompleteTime": { "hour": 0, "minute": 11, "second": 51.734 },
  "plotCompleteTime": { "hour": 7, "minute": 6, "second": 10.344 },
  "copyTime": { "hour": 0, "minute": 12, "second": 39.958 },
  "overallCompleteTime": { "hour": 7, "minute": 18, "second": 50.302 },
  "progress": 100,
  "phase": "complete",
  "finish_date": "2021-5-25 9:36:14"
}
```

## Donation
For continuous development, please support me with donation
`xch1wqpcvquv98qmeh9hg6wqpzhzmgs73lgvd8a7v5240nxgyat4p0sq4gdzyy`


CHANGELOG.md
## [0.0.3] - 2021-06-05
### Fixed
- Fixed an issue where it renders header unnecessarily every loop.
- Fixed an issue where it does not render at expected position when plot task count changes on `chkplot watch` command.

### Added
- Added no plot task message
- Added `-d` option to specify plot log directory.

## [0.0.2] - 2021-06-04
### Fixed
- Fixed an issue where new plot task was ignored in `chkplot watch` command

### Changed
- Changed monitor interval to 5s from 3s in `chkplot watch` command

## [0.0.1] - 2021-06-04
### Changed
- Updated CLI usage text

### Added
- Added new `chkplot watch` command
- Added gif demo movie to README
- Added `-c` option to `chkplot list` and `chkplot wip`

## [0.0.0] - 2021-06-02
Initial release.

<!-- [Unreleased]: https://github.com/Chia-Mine/chia-agent/compare/v0.0.1...v0.0.2 -->
[0.0.3]: https://github.com/Chia-Mine/plot-log-analyzer/compare/v0.0.2...v0.0.3
[0.0.2]: https://github.com/Chia-Mine/plot-log-analyzer/compare/v0.0.1...v0.0.2
[0.0.1]: https://github.com/Chia-Mine/plot-log-analyzer/compare/v0.0.0...v0.0.1
[0.0.0]: https://github.com/Chia-Mine/plot-log-analyzer/releases/tag/v0.0.0


src/file.ts
import * as path from "path";
import {promises as fs, Stats} from "fs";
import {homedir} from "os";
import {summarize} from "./analyze";

export async function listPlotterLogFiles(dirPath?: string){
  const CHIA_HOME = process.env.CHIA_HOME;
  let plotterDirPath = "";
  if(dirPath){
    plotterDirPath = path.resolve(dirPath);
  }
  else if(CHIA_HOME){
    plotterDirPath = path.resolve(CHIA_HOME, "plotter");
  }
  else{
    plotterDirPath = path.resolve(homedir(), ".chia", "mainnet", "plotter");
  }
  
  const files = await fs.readdir(plotterDirPath);
  const tasks: Promise<{ stats: Stats, path: string }>[] = [];
  for (const f of files){
    const p = path.resolve(plotterDirPath, f);
    tasks.push(fs.stat(p).then(s => ({stats: s, path: p})));
  }
  
  let fileStats = await Promise.all(tasks);
  fileStats = fileStats.filter(fs => {
    return fs.stats.isFile() && /plotter_log.*\.txt$/.test(path.basename(fs.path));
  });
  
  return fileStats;
}

export type TGetPlotterLogSummaryOption = {
  unfinishedOnly: boolean;
  n: number;
};
export async function* getPlotterLogSummary(
  fileStats: ReturnType<typeof listPlotterLogFiles> extends Promise<infer T> ? T : never,
  option?: Partial<TGetPlotterLogSummaryOption>,
){
  let files = [...fileStats];
  files.sort((a, b) => {
    return b.stats.birthtimeMs - a.stats.birthtimeMs;
  });
  
  const n = typeof option?.n === "number" ? option.n : -1;
  
  let i = 0;
  for(const f of files){
    let isYielded = false;
    const log = await fs.readFile(f.path, {encoding: "utf8"});
    const regex = /plotter_log_(.+)\.txt$/;
    const match = regex.exec(f.path);
    let uuid = "-";
    if(match){
      uuid = match[1];
    }
    const summary = summarize(log, uuid);
    if(option?.unfinishedOnly){
      if(summary.phase !== "complete"){
        yield summary;
        isYielded = true;
      }
    }
    else{
      yield summary;
      isYielded = true;
    }
    
    if(n > -1 && i >= n - 1){
      return;
    }
    
    if(isYielded){
      i++;
    }
  }
  
  return;
}


src/analyze.ts
import {parsePlotterLog, TParsed, TSize, TSizeUnit} from "./parse";

export function resolveSizeUnit(size: [number, string]){
  let p = 0;
  if(size[1] === "KiB"){
    p = 1;
  }
  if(size[1] === "MiB"){
    p = 2;
  }
  else if(size[1] === "GiB"){
    p = 3;
  }
  else if(size[1] === "TiB"){
    p = 4;
  }
  else if(size[1] === "PiB"){
    p = 5;
  }
  else if(size[1] === "EiB"){
    p = 6;
  }
  
  return size[0] * 1024**p;
}

export function summarizeSizeUnit(sizeInByte: number, unit?: TSizeUnit): TSize {
  let p = 0;
  if(unit === "B"){
    p = 0;
  }
  else if(unit === "KiB"){
    p = 1;
  }
  else if(unit === "MiB"){
    p = 2;
  }
  else if(unit === "GiB"){
    p = 3;
  }
  else if(unit === "TiB"){
    p = 4;
  }
  else if(unit === "PiB"){
    p = 5;
  }
  else{
    const log = Math.floor(Math.log2(sizeInByte) / 10);
    if(log === 1){ p = 1; unit = "KiB"; }
    else if(log === 2){ p = 2; unit = "MiB"; }
    else if(log === 3){ p = 3; unit = "GiB"; }
    else if(log === 4){ p = 4; unit = "TiB"; }
    else if(log === 5){ p = 5; unit = "PiB"; }
    else if(log === 6){ p = 6; unit = "EiB"; }
    else{
      return [sizeInByte, "B"];
    }
  }
  
  return [sizeInByte / (1024**p), unit];
}

const FINISHED_LOG_LINES = 2626; // 128
const FINISHED_LOG_LINES_64 = 1379; // 64
const FINISHED_LOG_LINES_32 = 754; // 32

export function getProgress(log: string){
  const lines = log.trim().split(/\r\n|\r|\n/).length;
  return lines > FINISHED_LOG_LINES ? 1 : lines / FINISHED_LOG_LINES;
}

export function getPhase1TimeSeries(parsedLog: TParsed<true>){
  return [
    parsedLog.phase1.table1.f1_complete.time,
    parsedLog.phase1.table2.complete.time,
    parsedLog.phase1.table3.complete.time,
    parsedLog.phase1.table4.complete.time,
    parsedLog.phase1.table5.complete.time,
    parsedLog.phase1.table6.complete.time,
    parsedLog.phase1.table7.complete.time,
  ];
}

export function getPhase1CPUSeries(parsedLog: TParsed<true>){
  return [
    parsedLog.phase1.table1.f1_complete.cpu,
    parsedLog.phase1.table2.complete.cpu,
    parsedLog.phase1.table3.complete.cpu,
    parsedLog.phase1.table4.complete.cpu,
    parsedLog.phase1.table5.complete.cpu,
    parsedLog.phase1.table6.complete.cpu,
    parsedLog.phase1.table7.complete.cpu,
  ];
}

export function getPhase2TimeSeries(parsedLog: TParsed<true>){
  return [
    parsedLog.phase2.table7.time,
    parsedLog.phase2.table6.time,
    parsedLog.phase2.table5.time,
    parsedLog.phase2.table4.time,
    parsedLog.phase2.table3.time,
    parsedLog.phase2.table2.time,
  ];
}

export function getPhase2CPUSeries(parsedLog: TParsed<true>){
  return [
    parsedLog.phase2.table7.cpu,
    parsedLog.phase2.table6.cpu,
    parsedLog.phase2.table5.cpu,
    parsedLog.phase2.table4.cpu,
    parsedLog.phase2.table3.cpu,
    parsedLog.phase2.table2.cpu,
  ];
}

export function parseTime(time: number){
  return {
    hour: Math.floor(time / 60 / 60),
    minute: Math.floor((time % 3600) / 60),
    second: Math.round((time % 60) * 1000) / 1000,
  };
}

export function compileTime(parsedTime: ReturnType<typeof parseTime>){
  return parsedTime.hour*3600 + parsedTime.minute*60 + parsedTime.second;
}

export function summarize(log: string, uuid?: string){
  const parsedLog = parsePlotterLog(log);
  const startDate = parsedLog.params.start_time || parsedLog.phase1.start_date;
  const endDate = parsedLog.phaseSummary.finished_time;
  
  const phase1CompleteTime = parsedLog.phase1.complete ? parseTime(parsedLog.phase1.complete.time) : null;
  const phase2CompleteTime = parsedLog.phase2.complete ? parseTime(parsedLog.phase2.complete.time) : null;
  const phase3CompleteTime = parsedLog.phase3.complete ? parseTime(parsedLog.phase3.complete.time) : null;
  const phase4CompleteTime = parsedLog.phase4.complete ? parseTime(parsedLog.phase4.complete.time) : null;
  const plotCompleteTime = parsedLog.phaseSummary.complete ? parseTime(parsedLog.phaseSummary.complete.time) : null;
  const copyTime = parsedLog.copyPhase.complete ? parseTime(parsedLog.copyPhase.complete.time) : null;
  const timeForPlotPlusCopy = parsedLog.phaseSummary.complete ? (parsedLog.phaseSummary.complete.time + (parsedLog.copyPhase.complete?.time || 0)) : null;
  const overallCompleteTime = timeForPlotPlusCopy ? parseTime(timeForPlotPlusCopy) : null;
  
  const phase = copyTime ? "complete" : plotCompleteTime ? "complete" : phase4CompleteTime ? "final copy"
  : phase3CompleteTime ? "phase4" : phase2CompleteTime ? "phase3" : phase1CompleteTime ? "phase2" : "phase1";
  
  return {
    uuid,
    id: parsedLog.params.id,
    start_date: startDate ? startDate : null,
    k: parsedLog.params.k,
    r: parsedLog.params.threads,
    b: parsedLog.params.buffer_size,
    t: parsedLog.params.temp1_dir,
    d: parsedLog.params.final_dir,
    phase1CompleteTime,
    phase2CompleteTime,
    phase3CompleteTime,
    phase4CompleteTime,
    plotCompleteTime,
    copyTime,
    overallCompleteTime,
    progress: Math.floor(getProgress(log)*1000)/10,
    phase,
    finish_date: endDate ? endDate.toLocaleString() : null,
  };
}

export function formatDate(d: Date){
  const year = d.getFullYear();
  const month = (d.getMonth()+1).toString().padStart(2, "0");
  const day = d.getDate().toString().padStart(2, "0");
  const hour = d.getHours().toString().padStart(2, "0");
  const min = d.getMinutes().toString().padStart(2, "0");
  const sec = d.getSeconds().toString().padStart(2, "0");
  return `${year}${month}${day}_${hour}${min}${sec}`;
}

export function printSummary(summary: ReturnType<typeof summarize>){
  let longestKeyName = "";
  for(const key in summary) {
    if (summary.hasOwnProperty(key)) {
      longestKeyName = longestKeyName.length < key.length ? key : longestKeyName;
    }
  }
    
  for(const key in summary){
    if(!summary.hasOwnProperty(key)){
      continue;
    }
    const val = summary[key as keyof typeof summary];
    
    const formattedKey = key.padEnd(longestKeyName.length + 1, " ");
    let formattedValue = "";
    if(typeof val === "string" || typeof val === "number"){
      formattedValue = `${val}`;
    }
    else if (Array.isArray(val)){
      formattedValue = `${val[0]}${val[1]}`;
    }
    else if(val instanceof Date){
      formattedValue = `${formatDate(val)}`;
    }
    else if(val && typeof val === "object"){
      formattedValue = `${val.hour.toString().padStart(2, "0")}:${val.minute.toString().padStart(2, "0")}:${val.minute.toString().padStart(2, "0")}`;
    }
    
    console.log(`${formattedKey}: ${formattedValue}`);
  }
}

export type TPrintProgressOption = {
  shortUUID: boolean;
  noTempDir: boolean;
  noFinalDir: boolean;
};
export function printProgress(summary: ReturnType<typeof summarize>, option?: Partial<TPrintProgressOption>){
  const output: string[] = [];
  
  if(option && option.shortUUID){
    const uuid = (summary.uuid||"").split("-")[0];
    output.push(uuid.padEnd(8, " "));
  }
  else{
    output.push(`${(summary.uuid||"").padEnd(36, " ")}`);
  }
  
  output.push(`${summary.progress}%`.padEnd(5, " "));
  output.push(`${summary.phase}`);
  output.push(`${summary.start_date ? formatDate(summary.start_date) : ""}`);
  output.push(`k:${summary.k}`);
  output.push(`r:${summary.r}`);
  output.push(`b:${summary.b?.join("")}`);
  if(!option || !option.noTempDir){
    output.push(`t:${summary.t}`);
  }
  if(!option || !option.noFinalDir){
    output.push(`d:${summary.d || "?"}`);
  }
  
  console.log(output.join(" "));
}
